torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1175,15 +1177,21 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
 void ProcessGroupNCCL::heartbeatMonitor() {
   uint64_t heartBeatCounter = 0ULL;
   while (true) {
-    auto heartbeat = heartbeat_;
-    if (heartbeat > heartBeatCounter) {
-      heartBeatCounter = heartbeat;
-      // Sleep for enough time so that heartBeat will at least increase.
-      std::this_thread::sleep_for(std::chrono::milliseconds(
-          kWatchdogThreadSleepMillis * heartbeatTimeoutInSec_));
+    std::unique_lock<std::mutex> lock(monitorMutex_);
```

Comment:
no locks!

Reply:
this is fine because this is only used by this thread so that we can wake up early.

Reply:
Unless we are ok with busy-waiting, like check the flag every 2 secs something like that.

Reply:
i think this approach with lock is ok actually. add a comment to warn people not to acquire it.  
more frequent sleep is probably also fine.

Reply:
sure, will add comment. I am not a big fan of busy waiting, so I will go with this lock approach.